### DISCUSSION TITLE
USB WiFi with Realtek RTL8811AU - Compatibility

### DISCUSSION THREAD

**Discussion by Special-Niewbie** [thefloppydriver/pc2drc#6] [Q&A] [answered]
Hello @thefloppydriver ,

In the meantime I go on other tests for the Kernel Patch problem, before I buy other hardware, I have a USB stick with Realtek RTL8811AU hardware. Is it possible to include or use this USB stick to connect with the Wii U PAD?

Here I attach the technical specifications with the driver site:

**Name:**
Realtek RTL8811AU Wireless LAN 802.11ac USB 2.0

**Linux drivers site:**
https://blog.abysm.org/2020/03/realtek-802-11ac-usb-wi-fi-linux-driver-installation/

**Picture:** 
![51GwEln1vwL _AC_SY355_](https://user-images.githubusercontent.com/64843123/156977259-64b5e5b6-2946-433d-994f-b47366a47d2f.jpg)

**Comment by thefloppydriver** [comment]
I've tried that exact adapter and unfortunately it isn't compatible. I'll be testing other adapters soon, but adapter compatibility is the exception instead of the rule. 

EDIT: I'm also not able to add compatibility in the future, as this is an issue involving the kernel, firmware, device, and drivers, only one of which (the kernel) I can somewhat mess around with.

**Comment by Stormy178** [comment]
Doing a quick search of "get_tsf()" for drivers coded into the main linux kernel turned up these results.
https://github.com/torvalds/linux/search?q=get_tsf%28%29
I don't really know much, but maybe you can pin this after sifting through that and making a quick list of compatible WiFi chipsets?

Taking a quick glance at it though, my current Wifi adapter(Netgear A6210/Mediatek MT7612U) is not compatible, well then, time to find a compatible one.

**Comment by Special-Niewbie** [comment]
I have another USB adapter WIFI and Bluetooth that I bought a year ago and called Comfast CF-726B, in case I will try this too.

Picture
![c453e88ecd22bd214232f8a9090e421d](https://user-images.githubusercontent.com/64843123/156985624-f5bfe471-3534-4159-9348-64659d0a718c.jpg)

**Comment by thefloppydriver** [comment]
@Stormy178 @Special-Niewbie A google search for "RT5572" under the shopping category will turn up quite a few compatible adapters. I'm trying to widen support for more adapters, but the feature that the gamepad requires is extremely niche and no adapter was ever required to support it to conform to the 802.11n standard. There are some new adapters that support it, but none with the kernel patch that I'm using currently.

**Comment by Special-Niewbie** [comment]
> @Stormy178 @Special-Niewbie A google search for "RT5572" under the shopping category will turn up quite a few compatible adapters. I'm trying to widen support for more adapters, but the feature that the gamepad requires is extremely niche and no adapter was ever required to support it to conform to the 802.11n standard. There are some new adapters that support it, but none with the kernel patch that I'm using currently.

Thanks for the answer of that, I don't know if the compatibility for the PAD it's extreimely niche because MAYBE the one that I have (and the one that you have tested Realtek RTL8811AU) it works the 802.11n standard under 2.4 Ghz and under 5 Ghz works only the AC protocols (that anyway include the 802.11n). MAYBE the PAD wants just the protocol 802.11n standard  under 5 Ghz.....

**Comment by thefloppydriver** [comment]
The gamepad requires 
   802.11n 5G compliance (so b/g/n/ac would work, as long as the device lists n compliance)
   The TSF hardware registers of the wireless adapter to be sent along with any video/audio data sent to the pad
     ^ That's where the problem is

In order to get that TSF value from the wireless adapter we have to patch the kernel. The kernel patch only works for devices which already support the get_tsf function and the mac80211 driver. It's difficult because very few devices support the get_tsf function, and there's no standard way to get the TSF value even with the few devices that support it.

You *can* connect to the pad without the kernel patch, in fact I've done it before on an Intel AX200 pci-E card. But you *cannot* send video at all without the kernel patch. I've tried every way around it, but both the kernel patch and a supported wifi card are required.

**Comment by Special-Niewbie** [comment]
> The gamepad requires 
>    802.11n 5G compliance (so b/g/n/ac would work, as long as the device lists n compliance)
>    The TSF hardware registers of the wireless adapter to be sent along with any video/audio data sent to the pad
>      ^ That's where the problem is
> 
> In order to get that TSF value from the wireless adapter we have to patch the kernel. The kernel patch only works for devices which already support the get_tsf function and the mac80211 driver. It's difficult because very few devices support the get_tsf function, and there's no standard way to get the TSF value even with the few devices that support it.
> 
> You *can* connect to the pad without the kernel patch, in fact I've done it before on an Intel AX200 pci-E card. But you *cannot* send video at all without the kernel patch. I've tried every way around it, but both the kernel patch and a supported wifi card are required.

Now I get......thank you. 

Stupid question, if you are using the Android phone by tethering wifi mode?

**Comment by thefloppydriver** [comment]
@Special-Niewbie Not unless you have a patch for the linux kernel on a rooted android device that has a compatible wifi chipset in it :joy:

**Comment by thefloppydriver** [comment]
@Special-Niewbie  
https://github.com/torvalds/linux/blob/8efd0d9c316af470377894a6a0f9ff63ce18c177/include/net/mac80211.h#L3637
```
3637 * @get_tsf: Get the current TSF timer value from firmware/hardware. Currently,
3638 *	this is only used for IBSS mode BSSID merging and debugging. Is not a
3639 *	required function.
3640 *	The callback can sleep.
```
^ mac80211.h
*this is not a required function*
The only 5GHz USB devices that support it are those that use ath9k, ath5k, carl9170, and rt2800usb.
There are modern pci-E devices that support it, such as mt7615 (b/g/n/ac) and mt7915 (a/b/g/n/ac/ax aka wifi 6) but those are costly (~70 USD) and I assume that you don't want to use a pci-E device.

**Comment by Special-Niewbie** [comment]
> @Special-Niewbie  
> https://github.com/torvalds/linux/blob/8efd0d9c316af470377894a6a0f9ff63ce18c177/include/net/mac80211.h#L3637
> ```
> 3637 * @get_tsf: Get the current TSF timer value from firmware/hardware. Currently,
> 3638 *	this is only used for IBSS mode BSSID merging and debugging. Is not a
> 3639 *	required function.
> 3640 *	The callback can sleep.
> ```
> ^ mac80211.h
> *this is not a required function*
> The only 5GHz USB devices that support it are those that use ath9k, ath5k, carl9170, and rt2800usb.
> There are modern pci-E devices that support it, such as mt7615 (b/g/n/ac) and mt7915 (a/b/g/n/ac/ax aka wifi 6) but those are costly (~70 USD) and I assume that you don't want to use a pci-E device.

Thank you so much for the info @thefloppydriver , so the hardware to looking for this project it needs to be in those listed chipset range..... Great, I will try to looking around, and share some models for double checking with you guys. 

No, the PCI-E WiFi i'm not looking for, also because don't give me the flexibility that the usb be can allow me to switch to a different PC/notebook....

**Comment by thefloppydriver** [comment]
@Special-Niewbie No problem! The only one Ive tested is the RT5572, so if you want something that's guaranteed to work I suggest getting something with that chipset

**Comment by Special-Niewbie** [comment]
> @Special-Niewbie No problem! The only one Ive tested is the RT5572, so if you want something that's guaranteed to work I suggest getting something with that chipset

Yesterday I was looking this model from ASUS that they advertising about the streaming video  as well ( I believe such product must to have the streaming WiFi functions)....

This it's the model:

Asus usb AC53 nano

![IMG_20220308_113545.jpg](https://user-images.githubusercontent.com/64843123/157161332-3f31b26b-060e-4b01-aa85-d712911dc910.jpg)

**Comment by Stormy178** [comment]
> > @Special-Niewbie No problem! The only one Ive tested is the RT5572, so if you want something that's guaranteed to work I suggest getting something with that chipset
> 
> Yesterday I was looking this model from ASUS that they advertising about the streaming video as well ( I believe such product must to have the streaming WiFi functions)....
> 
> This it's the model:
> 
> Asus usb AC53 nano
> 
> ![IMG_20220308_113545.jpg](https://user-images.githubusercontent.com/64843123/157161332-3f31b26b-060e-4b01-aa85-d712911dc910.jpg)

Using this site that I've found, looks like no drivers exist for it at all.
http://linux-hardware.org/?id=usb:0b05-17c9
Fortunately, their Find Parts page includes a section to search by USB ID, using lsusb.
Also has a handy part for looking up parts by Manufacturer and model.

**Comment by thefloppydriver** [comment]
> AC53 nano

@Special-Niewbie Unfortunately that won't work. The gamepad requires the special "get_tsf" function to be supported by the device's drivers and firmware, which has nothing to do with video streaming. I've scoured the entire linux kernel and a bunch of other repositories and the only devices that will work are the ones that I listed.

**Comment by thefloppydriver** [comment]
@Stormy178 it uses the rtl8822bu chipset which isn't supported by the linux kernel, but I've found a repository that claims to support the device here https://github.com/maccuaa/asus-ac53-rtl8822bu

But even with supported drivers, the device still won't work for streaming to the gamepad.

**Comment by Special-Niewbie** [comment]
@thefloppydriver and @Stormy178  seems I found a site with the list of all the chipset mentioned from @thefloppydriver ....

https://www.0xf8.org/2016/02/market-overview-of-usb-wifi-adapters-based-on-atheros-chipsets/

I would like just double confirm with you guys, it's this right products WIFI listed with the chipset that we need for this project?

If yes I will looking for one of them in my country....

**Comment by thefloppydriver** [comment]
> @thefloppydriver and @Stormy178  seems I found a site with the list of all the chipset mentioned from @thefloppydriver ....
> 
> https://www.0xf8.org/2016/02/market-overview-of-usb-wifi-adapters-based-on-atheros-chipsets/
> 
> I would like just double confirm with you guys, it's this right products WIFI listed with the chipset that we need for this project?
> 
> If yes I will looking for one of them in my country....

@Special-Niewbie Yep, those are compatible!

**Comment by Special-Niewbie** [comment]
Okay, after a bit sourcing of which models I like and also the availability of them in my area; I have found the below models:

1) **D-Link DWA-130**

![DWA 130](https://user-images.githubusercontent.com/64843123/157200762-ae9e205e-434a-4cec-a918-2de53b52bbc4.jpg)


2) **TL-WN722N** (the one that I found it's the TL-WN722NC , the "C" it's for USB Cradle (with the base) )

![TL-WN722N](https://user-images.githubusercontent.com/64843123/157201065-17640b2a-c5d3-42b8-9f92-f678a5ddddd0.jpg)


For good Gain, I was thinking to take the **TL-WN722NC**, what do you think?

EDIT 1: I don't think to buy the TL-WN722NC because i noticed that it's 150 Mbps and not 300 Mbps.... I need to search some model more higher

EDIT 2: So far the best model it seems the **D-Link DWA-160** better then the 130 model as well....

**Comment by Special-Niewbie** [comment]
@thefloppydriver , okay after a lot studying of which is the best WiFi USB adapter, in the end seems the one that you picked up, the rt5572 (the same specs as the D-Link DWA-160 that it's so rare to find in our days)....

All the rest of the models , ( as we know ) are really old performance, but to ketch the best performance from these old models seems that the rt5572 (not all models) it's the unique one that can deliver a maximum of performance under 2.4Ghz and 5Ghz respectively 300Mbit (or 600 in dual mode, that anyway what we are interested it's to focus the 5Ghz performance side).

Now, also yesterday I was thinking that after this , will be another obstacle for us. The obstacle it's, okay that the pad can receive the input of the screen by the WiFi 5Ghz, but the rest of the input are managing by the Bluetooth side....So, my question it's, we will face the same problem on Bluetooth side as the Wifi side? I mean we need a specific Bluetooth type of hardware as well?

**Comment by thefloppydriver** [comment]
@Special-Niewbie Actually, the gamepad has no Bluetooth module. Everything is handled over wifi.

**Comment by Special-Niewbie** [comment]
> @Special-Niewbie Actually, the gamepad has no Bluetooth module. Everything is handled over wifi.

No, because  one period I was working on the software of Nintendont for Wii u, and the input of the joystick it's management by Bluetooth and the screen by the 5Ghz......

**Comment by thefloppydriver** [comment]
> > @Special-Niewbie Actually, the gamepad has no Bluetooth module. Everything is handled over wifi.
> 
> No, because  one period I was working on the software of Nintendont for Wii u, and the input of the joystick it's management by Bluetooth and the screen by the 5Ghz......

I promise it's all over wifi, I have screenshots of the input events actually coming over the WiFi interface viewed through Wireshark. I can show them to you tomorrow if you'd like to see them. The WiiU Pro Controller and Wiimotes both use Bluetooth, however the gamepad does *all* communication over wifi.

**Comment by thefloppydriver** [comment, marked as answer]
> Okay, after a bit sourcing of which models I like and also the availability of them in my area; I have found the below models:
> 
> 1. **D-Link DWA-130**
> 
> ![DWA 130](https://user-images.githubusercontent.com/64843123/157200762-ae9e205e-434a-4cec-a918-2de53b52bbc4.jpg)
> 
> 2. **TL-WN722N** (the one that I found it's the TL-WN722NC , the "C" it's for USB Cradle (with the base) )
> 
> ![TL-WN722N](https://user-images.githubusercontent.com/64843123/157201065-17640b2a-c5d3-42b8-9f92-f678a5ddddd0.jpg)
> 
> For good Gain, I was thinking to take the **TL-WN722NC**, what do you think?
> 
> EDIT 1: I don't think to buy the TL-WN722NC because i noticed that it's 150 Mbps and not 300 Mbps.... I need to search some model more higher
> 
> EDIT 2: So far the best model it seems the **D-Link DWA-160** better then the 130 model as well....

@Special-Niewbie 
I'm sorry, I didn't see these. I wouldn't use either of them because the D-Link DWA-130 doesn't support the wifi functionality required for this project, and the TL-WN722N (and TL-WN722NC) have many different, unlabeled revisions and only the first two (oldest, v1.0 and v1.1) are supported. The newer (v2 and v3) revisions use the rtl8188eus chipset which does not support the wifi functionality required for this project.

I suggest going here https://www.google.com/search?q=rt5572&tbm=shop and seeing what you can find.

**Comment by Special-Niewbie** [comment]
> > Okay, after a bit sourcing of which models I like and also the availability of them in my area; I have found the below models:
> > 
> > 1. **D-Link DWA-130**
> > 
> > ![DWA 130](https://user-images.githubusercontent.com/64843123/157200762-ae9e205e-434a-4cec-a918-2de53b52bbc4.jpg)
> > 
> > 2. **TL-WN722N** (the one that I found it's the TL-WN722NC , the "C" it's for USB Cradle (with the base) )
> > 
> > ![TL-WN722N](https://user-images.githubusercontent.com/64843123/157201065-17640b2a-c5d3-42b8-9f92-f678a5ddddd0.jpg)
> > 
> > For good Gain, I was thinking to take the **TL-WN722NC**, what do you think?
> > 
> > EDIT 1: I don't think to buy the TL-WN722NC because i noticed that it's 150 Mbps and not 300 Mbps.... I need to search some model more higher
> > 
> > EDIT 2: So far the best model it seems the **D-Link DWA-160** better then the 130 model as well....
> 
> @Special-Niewbie 
> I'm sorry, I didn't see these. I wouldn't use either of them because the D-Link DWA-130 doesn't support the wifi functionality required for this project, and the TL-WN722N (and TL-WN722NC) have many different, unlabeled revisions and only the first two (oldest, v1.0 and v1.1) are supported. The newer (v2 and v3) revisions use the rtl8188eus chipset which does not support the wifi functionality required for this project.
> 
> I suggest going here https://www.google.com/search?q=rt5572&tbm=shop and seeing what you can find.

Thank you to reply on that, but no worries about because these results was on the beginning of my research , except the revisions of the models that I was meticulous to find before buying , but also these models doesn't support or the 5 GHz or the transfer rate it's 150 Mbps instead of 300 ...

Only the DWA- 160 was good to consider, that the unbranded models like you bought, have the same hardware base as the DWA 160 . So in our days in Asia it's possible to find the DWA 160 hardware in all specific unbranded products and it's good news about; otherwise these pieces of hardware are difficult to find around.

**Reply by thefloppydriver** [reply]
@Special-Niewbie The D-Link DWA-160 uses the AR9170 chipset which uses the carl9170 Linux drivers which are supported! I would like to add that the device that I have and the D-Link DWA-160 do not use the same chipset. Mine has the rt5572 chipset which uses the rt2x00usb drivers :)

**Reply by Special-Niewbie** [reply]
> @Special-Niewbie The D-Link DWA-160 uses the AR9170 chipset which uses the carl9170 Linux drivers which are supported! I would like to add that the device that I have and the D-Link DWA-160 do not use the same chipset. Mine has the rt5572 chipset which uses the rt2x00usb drivers :)

I talking about the newest/latest revision of DWA-160 "C1" (the B2 as well) has the same rt5572 like yours, also here this model has so different many revisions..... ;)

**Reply by thefloppydriver** [reply]
@Special-Niewbie I see! It's perfectly fine then!

**Reply by Special-Niewbie** [reply]
@thefloppydriver , after try to find in my area some of the models mentioned in our previous posts ; well it's incredible hard to find any models. So, today, I decided to buy the USB WiFi by internet and I bought the following model:

![88883006.jpg](https://user-images.githubusercontent.com/64843123/158027100-74cea5fe-a778-41fe-b5d4-6ea18770a405.jpg)

I bought this model , because it seems quite strong, and I tried to looking something different from your because at least we have different models to tests.... This Netis model it has the same chipset like your the RT5572. So , as soon I will receive it I will let you know. Keep in touch

**Reply by thefloppydriver** [reply]
@Special-Niewbie Great idea! I'll be waiting!

**Comment by Special-Niewbie** [comment]
@thefloppydriver , I hope you are going well..... 

Around next hours / days I should receive the wireless Pen Drive , it's on the way to my home :)...

**Reply by thefloppydriver** [reply]
Can't wait for it to get there! Tell me how it goes!

**Reply by Special-Niewbie** [reply]
Okay, after a bit delayed from the postage.... finally I got my parcel, tomorrow I will continue with the testing ....

**Reply by Special-Niewbie** [reply]
Because I cannot wait , after so much waiting, I tried to connect the USB and everything it works perfectly. I have tried to build but give me the error....now i restarted the OS and re-try the procedure..
